lib/object.gi

```diff
@@ -763,27 +763,44 @@ InstallGlobalFunction( MU_Finalize, function (  )
   fi;
 end );
 
-InstallMethod( MemoryUsage, "fallback method for objs without subobjs",
+InstallMethod( MemoryUsage, "generic fallback method",
   [ IsObject ],
   function( o )
-    local mem;
-    mem := SHALLOW_SIZE(o);
-    if mem = 0 then 
+    local mem,known,i,s;
+
+    if SHALLOW_SIZE(o) = 0 then
         return MU_MemPointer;
+    fi;
+
+    if MU_AddToCache( o ) then
+        return 0;    # already counted
+    fi;
+
+    MEMUSAGECACHE.depth := MEMUSAGECACHE.depth + 1;
+
+    # Count the bag, the header, and the master pointer
+    mem := SHALLOW_SIZE(o) + MU_MemBagHeader + MU_MemPointer;
+    if IS_POSOBJ(o) then
+        # Again the bag, its header, and the master pointer
+        for i in [1..LEN_POSOBJ(o)] do
+            if IsBound(o![i]) then
+                if SHALLOW_SIZE(o![i]) > 0 then    # a subobject!
+                    mem := mem + MemoryUsage(o![i]);
+                fi;
+            fi;
+        od;
+    elif IS_COMOBJ(o) then
+        # Again the bag, its header, and the master pointer
+        for i in NamesOfComponents(o) do
+            s := o!.(i);
+            if SHALLOW_SIZE(s) > 0 then    # a subobject!
+                mem := mem + MemoryUsage(s);
+            fi;
+        od;
     else
-        # a proper object, thus we have to add it to the database
-        # to not count it again!
-        if not(MU_AddToCache(o)) then
-            mem := mem + MU_MemBagHeader + MU_MemPointer;
-            # This is for the bag, the header, and the master pointer
-        else 
-            mem := 0;   # already counted
-        fi;
-        if MEMUSAGECACHE.depth = 0 then   
-            # we were the first to be called, thus we have to do the cleanup
-            ClearObjectMarker( MEMUSAGECACHE );
-        fi;
+        # objs without subobjs
```

Comment:
I added the warning, but I had to move `MemoryUsage` to a new file, as `InfoWarning` is not available in `object.gi`. But I actually think this is better anyway.